sdks/python/apache_beam/dataframe/transforms.py

```diff
@@ -0,0 +1,244 @@
+      self,
+      inputs,  # type: Dict[PlaceholderExpr, PCollection]
+      outputs,  # type: Dict[Any, Expression]
+      ):  # -> Dict[Any, PCollection]
+    """Construct a Beam graph that evaluates a set of expressions on a set of
+    input PCollections.
+
+    :param inputs: A mapping of placeholder expressions to PCollections.
+    :param outputs: A mapping of keys to expressions defined in terms of the
+        placeholders of inputs.
+
+    Returns a dictionary whose keys are those of outputs, and whose values are
+    PCollections corresponding to the values of outputs evaluated at the
+    values of inputs.
+
+    Logically, `_apply_deferred_ops({x: a, y: b}, {f: F(x, y), g: G(x, y)})`
+    returns `{f: F(a, b), g: G(a, b)}`.
+    """
+    class ComputeStage(beam.PTransform):
+      """A helper transform that computes a single stage of operations.
+      """
+      def __init__(self, stage):
+        self.stage = stage
+
+      def default_label(self):
+        return '%s:%s' % (self.stage.ops, id(self))
+
+      def expand(self, pcolls):
+        if self.stage.is_grouping:
+          # Arrange such that partitioned_pcoll is properly partitioned.
+          input_pcolls = {
+              k: pcoll | 'Flat%s' % k >> beam.FlatMap(_partition_by_index)
+              for k,
+              pcoll in pcolls.items()
+          }
+          partitioned_pcoll = input_pcolls | beam.CoGroupByKey(
+          ) | beam.MapTuple(
+              lambda _, inputs: {k: pd.concat(vs)
+                                 for k, vs in inputs.items()})
+        else:
+          # Already partitioned, or no partitioning needed.
+          (k, pcoll), = pcolls.items()
+          partitioned_pcoll = pcoll | beam.Map(lambda df: {k: df})
+
+        # Actually evaluate the expressions.
+        def evaluate(partition, stage=self.stage):
+          session = expressions.Session(
+              {expr: partition[expr._id]
+               for expr in stage.inputs})
+          for expr in stage.outputs:
+            yield beam.pvalue.TaggedOutput(expr._id, expr.evaluate_at(session))
+
+        return partitioned_pcoll | beam.FlatMap(evaluate).with_outputs()
+
+    class Stage(object):
+      """Used to build up a set of operations that can be fused together.
+      """
+      def __init__(self, inputs, is_grouping):
+        self.inputs = set(inputs)
+        self.is_grouping = is_grouping or len(self.inputs) > 1
+        self.ops = []
+        self.outputs = set()
+
+    # First define some helper functions.
+    def output_is_partitioned_by_index(expr, stage):
+      if expr in stage.inputs:
+        return stage.is_grouping
+      elif expr.preserves_partition_by_index():
+        if expr.requires_partition_by_index():
+          return True
+        else:
+          return all(
+              output_is_partitioned_by_index(arg, stage) for arg in expr.args())
+      else:
+        return False
+
+    def _partition_by_index(df, levels=None, parts=10):
```

Comment:
nit: none of the other functions defined here have an underscore prefix, can you drop this one?